hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpContractStoresSubcommand.java

```diff
@@ -167,28 +196,125 @@ public static ContractKeyLocal from(ContractKey ckey) {
+                nSlotsSeen.get(), didRunToCompletion);
+
         return didRunToCompletion;
     }
 
+    boolean iterateThroughMigratedContractStorage(BiConsumer<ContractKeyLocal, UInt256> visitor) {
+
+        System.err.printf("iterateThroughMigratedContractStorage: start%n");
+
+        final var contractStorageStore = getMigratedContractStore();
+
+        System.err.printf("iterateThroughMigratedContractStorage: migrated%n");
+
+        final var nSlotsSeen = new AtomicLong();
+
+        if (contractStorageStore == null) return false;
+        contractStorageStore.keys().forEachRemaining(key -> {
+            // (Not sure how many temporary _copies_ of a byte arrays are made here ... best not to ask ...)
+            final var contractKeyLocal = ContractKeyLocal.from(
+                    new ContractKey(key.contractNumber(), key.key().toByteArray()));
+            final var slotValue = contractStorageStore.get(key);
+            assert (slotValue != null);
+            final var value = uint256FromByteArray(slotValue.value().toByteArray());
+            nSlotsSeen.incrementAndGet();
+            visitor.accept(contractKeyLocal, value);
+        });
+
+        System.err.printf("iterateThroughContractStorage: end - %d slots seen, completed? true%n", nSlotsSeen.get());
+
+        return true;
+    }
+
+    /** First migrates the contract store from the mono-service representation to the modular-service representations,
+     *  and then returns all contracts with bytecodes from the migrated contract store, plus the ids of contracts with
+     *  0-length bytecodes.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    WritableKVState<SlotKey, SlotValue> getMigratedContractStore() {
+
+        final var fromStore = state.getRawContractStorage();
+        final var expectedNumberOfSlots = Math.toIntExact(fromStore.size());
+
+        // Start the migration with a clean, writable KV store.  Using the in-memory store here.
+
+        final var contractSchema = new ContractSchema();
+        final var contractSchemas = contractSchema.statesToCreate();
+        final StateDefinition<SlotKey, SlotValue> contractStoreStateDefinition = contractSchemas.stream()
+                .filter(sd -> sd.stateKey().equals(ContractSchema.STORAGE_KEY))
+                .findFirst()
+                .orElseThrow();
+        final var contractStoreSchemaMetadata =
+                new StateMetadata<>(ContractService.NAME, contractSchema, contractStoreStateDefinition);
+        final var contractMerkleMap =
+                new NonAtomicReference<MerkleMap<InMemoryKey<SlotKey>, InMemoryValue<SlotKey, SlotValue>>>(
+                        new MerkleMap<>(expectedNumberOfSlots));
+        final var toStore = new NonAtomicReference<WritableKVState<SlotKey, SlotValue>>(
+                new InMemoryWritableKVState<>(contractStoreSchemaMetadata, contractMerkleMap.get()));
```

Comment:
My initial reaction to this is that it feels like we're pulling out the guts of the Schema here, effectively ignoring a piece of the abstraction instead of working with it. Put another way, does the Schema provide a straightforward way (or could we add one) such that these details could be appropriately encapsulated? Or is that too much work for now? Not sure 🤔

Reply:
You're probably right that there's a better way to do it - but I don't know it.  I'll ask someone.  My goal here is just to get a proper `MerkleMap` and then wrap a `WriteableKVState` around it.  And, this is from the CLI, so I don't positiviely need to do it any "official" way.  Just any random way will suffice, here.